[WIP] Add new handwriting recognition database 'IAM'

aarora8 opened the pull request:
This branch is for handwritten word recognition on text line images. It uses word language model.

add scripts for data preparation (text, wav.scp and utt2spk file) (local/prepare_data.sh, local/process_data.py)
add scripts for feature extraction (local/make_feature_vect.py)
add scripts for lexicon, language modeling, grammar (egs/iam/s5/local/prepare_lm.sh, egs/iam/s5/local/prepare_lexicon.py, egs/iam/s5/local/prepare_dict.sh)
add script for GMM-HMM training and using chain model (egs/iam/s5/local/chain/run_cnn_1a.sh, egs/iam/s5/local/chain/align_nnet3_lats.sh, egs/iam/s5/run.sh, egs/iam/s5/local/chain/run_cnn_chainali_1a.sh)
@ChunChiehChang @YiwenShaoStephen

hhadian commented on egs/iam/s5/local/chain/run_cnn_1b.sh:
```diff
@@ -0,0 +1,231 @@
+  relu-batchnorm-layer name=tdnn1 input=Append(-1,0,1) dim=$tdnn_dim
+  relu-batchnorm-layer name=tdnn2 input=Append(-4,0,4) dim=$tdnn_dim
+  relu-batchnorm-layer name=tdnn3 input=Append(-4,0,4) dim=$tdnn_dim
+  relu-batchnorm-layer name=tdnn4 input=Append(-4,0,4) dim=$tdnn_dim
```
Please either use -5,0,5 (and make other changes accordingly) or change frame subsampling factor to 4.
Also, please fix the time/height asymmetry issue that Dan pointed out (i.e. [for frame-subsmple-factor of 4] use splicing -4,-2,0,2,4 for cnn3,4,5,6 and tdnn1)

hhadian commented on egs/iam/s5/local/chain/run_cnn_1b.sh:
```diff
@@ -0,0 +1,231 @@
+gmm=tri2_8states_4sil_10000_500_20000_500_20000        # this is the source gmm-dnir that we'll use for alignments; it
+
+                # should have alignments for the specified training data.
+nnet3_affix=_cleaned    # affix for exp dirs, e.g. it was _cleaned in tedlium.
```
Why are you using a `_cleaned` affix?

hhadian commented on egs/iam/s5/local/chain/run_cnn_1b.sh:
```diff
@@ -0,0 +1,231 @@
+
+nj=30
+train_set=train
+gmm=tri2_8states_4sil_10000_500_20000_500_20000        # this is the source gmm-dnir that we'll use for alignments; it
```
This seems informative and good, but I guess it's best if we simply change it to tri2.
You can mention in a comment in `run.sh` that the best gmm has these configurations.
BTW, if there is no `1a`, change this to `1a` (instead of `1b`).
Also, please add a stage for running the chain experiments in `run.sh`
Please fix the typo in the comment and the extra new line.

hhadian commented on egs/iam/s5/local/nnet3/compare.sh:
```diff
@@ -0,0 +1,56 @@
+#!/bin/bash
```
You forgot to fix and rename this script.

hhadian commented on egs/iam/s5/local/add_test_data_to_train.py:
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env python
```
This script is not needed really. 
You can add a few lines to run.sh to do this:
Get the set of unique words in test text by `cat data/test/text | awk '{ for(i=2;i<=NF;i++) print $i;}' | sort -u >test_words.txt`. Do this for train text too.
Subtract train words from test words: `filter_scp.pl --exclude train_words.txt test_words.txt >diff.txt`
Append them to train text: `cat diff.txt | awk '{ print "id " $1 }' >> train/text`

Actually, we won't need this type of LM altogether after we adopt a standard LM like the one in that paper you mentioned.

hhadian commented on egs/iam/s5/local/score.sh:
```diff
@@ -0,0 +1,149 @@
+#!/bin/bash
```
Make this a link to steps/scoring/score_kaldi_wer.sh

hhadian commented on egs/wsj/s5/steps/nnet3/decode.sh:
```diff
@@ -163,7 +163,7 @@ if [ $stage -le 3 ]; then
       echo "Not scoring because local/score.sh does not exist or not executable." && exit 1;
```
revert the changes in this file

aarora8 commented:
- updated run_cnn scripts, added stage for running chain models in run.sh, updated compare_wer.sh, removed add_test_data_to_train.py
- testing code with symbolic link for score.sh, reverting change in decode.sh

hhadian commented on egs/iam/s5/run.sh:
```diff
@@ -0,0 +1,175 @@
+  cat $data_dir/test/text | awk '{ for(i=2;i<=NF;i++) print $i;}' | sort -u >test_words.txt
+  cat $data_dir/train/text | awk '{ for(i=2;i<=NF;i++) print $i;}' | sort -u >train_words.txt
+  filter_scp.pl --exclude train_words.txt test_words.txt >diff.txt
+  cat diff.txt | awk '{ print "id " $1 }' >> $data_dir/train/text
```
You should make a copy of train/text first. You don't want to affect training.

hhadian commented on egs/wsj/s5/steps/nnet3/decode.sh:
```diff
@@ -163,7 +163,7 @@ if [ $stage -le 3 ]; then
     echo "score best paths"
     [ "$iter" != "final" ] && iter_opt="--iter $iter"
-    local/score.sh $iter_opt $scoring_opts --cmd "$cmd" $data $graphdir $dir
+    local/score.sh $scoring_opts --cmd "$cmd" $data $graphdir $dir
```
You have not fixed this yet.

aarora8 commented:
- updated run script for creating copy of training data
- reverted changes in decode script

hhadian commented on egs/iam/s5/local/chain/run_cnn_chainali_1a.sh:
```diff
@@ -0,0 +1,238 @@
+  conv-relu-batchnorm-layer name=cnn3 height-in=20 height-out=20 time-offsets=-4,-2,0,2,4 $common2
+  conv-relu-batchnorm-layer name=cnn4 height-in=20 height-out=10 time-offsets=-4,-2,0,2,4 $common2 height-subsample-out=2
+  conv-relu-batchnorm-layer name=cnn5 height-in=10 height-out=10 time-offsets=-4,0,4 $common3
+  conv-relu-batchnorm-layer name=cnn6 height-in=10 height-out=10 time-offsets=-4,0,4 $common3
```
I didn't notice you added more layers. What is the WER w/o extra layers?
It would be good to make 1a and chain_1a the same and only change the source of alignments.

hhadian replied:
Ashish, remove these extra layers and make it like 1a. We can add them in future recipes if they are helpful (e.g. 1b, 1c, etc.).

aarora8 replied:
sorry for the delay. updated configuration, will surely add proper word error rate value today.

hhadian commented on egs/iam/s5/local/chain/compare_wer.sh:
```diff
@@ -0,0 +1,63 @@
+for x in $*; do   printf "% 10s" " $(basename $x)";   done
+echo
+
+for x in $*; do
```
Add a `echo -n "# WER            "` before this `for`.
Also, please tune the number of spaces so that all numbers are vertically aligned (in a "fixed width" font like in terminal) and then update the example at the top of the file.

hhadian commented on egs/iam/s5/local/chain/compare_wer.sh:
```diff
@@ -0,0 +1,63 @@
+echo
+
+for x in $*; do
+  wer=$(cat $x/decode_test/scoring_kaldi/best_wer | utils/best_wer.sh | awk '{print $2}')
```
I guess you don't need to pass it through utils/best_wer.sh

YiwenShaoStephen commented on egs/iam/s5/run.sh:
```diff
@@ -0,0 +1,178 @@
+nnet3_affix=fsf4
+common_egs_dir=$exp_dir/chainfsf4/cnn1a_1/egs
+
+if [ $stage -le 130 ]; then
```
I think you might want to refer to "13" instead of "130" here?

aarora8 commented:
updated compare_wer.sh and run.sh. 
running run_cnn_chainali_1a.sh to get word error rate.

aarora8 commented:
added modification for removing affix from run.sh, output from compare_wer.sh. adding comments in  code.

hhadian commented on egs/iam/s5/local/chain/run_cnn_1a.sh:
```diff
@@ -0,0 +1,231 @@
+# exp/chainfsf4/cnn1a_1/: num-iters=21 nj=2..4 num-params=4.4M dim=40->380 combine=-0.033->-0.025 xent:train/valid[13,20,final]=(-1.07,-1.31,-0.560/-1.30,-1.70,-0.978) logprob:train/valid[13,20,final]=(-0.064,-0.119,-0.011/-0.115,-0.208,-0.096)
+
+# head exp/chainfsf4/cnn1a_1/decode_test/scoring_kaldi/best_wer
+#%WER 18.34 [ 3231 / 17616, 348 ins, 693 del, 2190 sub ] exp/chainfsf4/cnn1a_1/decode_test/wer_8_1.0
```
Please fix the directory names in these results (i.e. chain chainfsf4 to chain and cnn1a_1 to cnn1a)
BTW, we will later need to update them when you get the LM working so you can do it then.

hhadian commented on egs/iam/s5/local/chain/run_cnn_chainali_1a.sh:
```diff
@@ -0,0 +1,236 @@
+gmm=tri3        # this is the source gmm-dir that we'll use for alignments; it
+                # should have alignments for the specified training data.
+nnet3_affix=    # affix for exp dirs, e.g. it was _cleaned in tedlium.
+affix=1a_chainali  #affix for TDNN+LSTM directory e.g. "1a" or "1b", in case we change the configuration.
```
It's better to change this to `_1a`  and instead move the `chainali` 
affix to the `dir=...` line below (so the final dir name becomes `cnn_chainali_1a`) 
since that's the way it's done in other places in Kaldi.
Obviously update the names in the results too.

aarora8 committed:
OCR: Add IAM corpus with unk decoding support
